### PR TITLE
Feature message read status

### PR DESCRIPTION
**Pull request by T-mclennan**
Greetings! The goal of this feature branch was to add read status message notifications outlined in issue #2.

My approach was to add an "unseenCount" data member the Conversation object, and just keep track of the number of unseen messages contained:

- When a conversation is created unseenCount starts at 0, and each message added adds 1. (because at the point a message is first processed by the server, it is always unseen by recipient. ) 
- Incoming messages through the web-socket increment unseenCount if this conversation is not currently selected. 
- When a conversation is selected for viewing, this unseenCount is returned to 0 in both redux and the database. 
- When a conversation is fetched from the database by a user, a check is performed to see if this user is the unseen recipient, if not set unseenCount to 0.
- Badge component is used to display notifications to the user.

Here is a snapshot of the notifications:
Screen Shot 2021-07-16 at 11 34 11 AM

Thanks!

### CODE REVIEW

**Review by 0916dhkim — COMMENTED** [review]
@T-mclennan Good job overall. The new docstrings for API route handlers are excellent :+1: 
Here are things we can improve in this PR:
1. On the provided design specs, you can find a small avatar below the last message read by the other user. Please add it.
2. When a conversation is already open on screen, any new incoming message should be automatically marked as seen.
3. I made a few minor nitpicks here and there.

Let the reviewers know when this PR is ready for another review!

File: server/routes/api/conversations.js [hunk 1]
```diff
@@ -1,22 +1,56 @@
...
+ * @param {callback} middleware - Express middleware.
+ * @returns {object} 200 - success
+ * @returns {Error}  401 - Validation error
+ */
+router.post("/viewed/:id", async (req, res, next) => {
+  try {
+    if (!req.user) {
+      return res.sendStatus(401);
+    }
+    const updatedConvo = await Conversation.resetUnseenCount(req.params.id)
```
[the remark this answers is not in the document]
Comment (T-mclennan) [reply server/routes/api/conversations.js]: Okay I added some verification logic to the route handler, let me know if that's a sufficient way to address the issue.

File: client/src/reducers/activeConversationUserId.js [hunk 2]
```diff
@@ -1,9 +1,9 @@
 import {SET_ACTIVE_CHAT} from '../actions/actionTypes'
 
-const activeConversationReducer = (state = "", action) => {
+const activeConversationReducer = (state = null, action) => {
   switch (action.type) {
     case SET_ACTIVE_CHAT: {
-      return action.username;
+      return action.userId;
```
[the remark this answers is not in the document]
Comment (T-mclennan) [reply client/src/reducers/activeConversationUserId.js:6]: Thanks, I didn't notice that avatar! I will make those adjustments now. 

For point #2: This case is currently being handled in the socket logic, in src/socket.js. And actually it's the purpose for the code in the comment below this one where activeConversation should be activeConversationUserId.

[the remark this answers is not in the document]
Comment (T-mclennan) [reply client/src/reducers/activeConversationUserId.js:6]: Okay I added an avatar read receipt below the appropriate message.

File: server/db/models/conversation.js [hunk 3]
```diff
@@ -23,16 +35,73 @@ Conversation.findConversationById = async function (conversationId) {
...
+Conversation.createConversation = async function (senderId, recipientId ) {
   if (isNaN(senderId) || isNaN(recipientId)) {
     throw new Error("Error: Conversation could not be created. Two valid user Id's are needed.")
   }
   conversation = await Conversation.create({
     user1Id: senderId,
     user2Id: recipientId,
+    lastMessageReadUser1: null,
+    lastMessageReadUser2: null,
+    unseenCount: 0,
```
Comment (0916dhkim) [inline server/db/models/conversation.js:47]: If we know the last read message of each user, `unseenCount` can be calculated. We do not need to persist this value in database.

**Review by 0916dhkim — APPROVED** [review]
@T-mclennan Great job. I see the avatar is added & messages are marked as seen in real-time. This PR is approved :100: 
You can fix the following glitches before merging:
- Opening a new conversation (by sending the first message to a user) crashes the application. Try texting Julia as Santiago.
- Avatar does not update real-time.

File: server/db/models/conversation.js (hunk 3 above)
Comment (T-mclennan) [reply server/db/models/conversation.js:47]: Excellent, I will investigate those issues and fix them before merging. That's a great insight on using the last read message to calculate the unseenCount value. Thank you!

**Pull request merged**